core/lib/zksync_core/src/eth_watch/event_processors/governance_upgrades.rs

```diff
@@ -0,0 +1,117 @@
+use crate::eth_watch::{
+    client::{Error, EthClient},
+    event_processors::EventProcessor,
+};
+use std::convert::TryFrom;
+use std::time::Instant;
+use zksync_contracts::governance_contract;
+use zksync_dal::StorageProcessor;
+use zksync_types::{
+    protocol_version::GovernanceOperation, web3::types::Log, Address, ProtocolUpgrade,
+    ProtocolVersionId, H256,
+};
+
+/// Responsible for saving new protocol upgrade proposals to the database.
+#[derive(Debug)]
+pub struct GovernanceUpgradesEventProcessor {
+    diamond_proxy_address: Address,
+    last_seen_version_id: ProtocolVersionId,
+    upgrade_proposal_signature: H256,
+}
+
+impl GovernanceUpgradesEventProcessor {
+    pub fn new(
+        diamond_proxy_address: Address,
+        last_seen_version_id: ProtocolVersionId,
+    ) -> Option<Self> {
+        let Some(governance_contract) = governance_contract() else {
+            return None;
+        };
+
+        Some(Self {
+            diamond_proxy_address,
+            last_seen_version_id,
+            upgrade_proposal_signature: governance_contract
+                .event("TransparentOperationScheduled")
+                .expect("TransparentOperationScheduled event is missing in abi")
+                .signature(),
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl<W: EthClient + Sync> EventProcessor<W> for GovernanceUpgradesEventProcessor {
+    async fn process_events(
+        &mut self,
+        storage: &mut StorageProcessor<'_>,
+        client: &W,
+        events: Vec<Log>,
+    ) -> Result<(), Error> {
+        let mut upgrades = Vec::new();
+        for event in events
+            .into_iter()
+            .filter(|event| event.topics[0] == self.upgrade_proposal_signature)
+        {
+            let governance_operation = GovernanceOperation::try_from(event)
+                .map_err(|err| Error::LogParse(format!("{:?}", err)))?;
+            // Some calls can target other contracts than Diamond proxy, skip them.
+            for call in governance_operation
```

Comment:
Is it possible to have more than 1 upgrade?

Reply:
It's allowed, but we won't do so I suppose. There can be `calls` that are no upgrade I think. @vladbochok can explain better